src/librustc_mir/transform/const_prop.rs

```diff
@@ -988,8 +1032,53 @@ impl<'mir, 'tcx> MutVisitor<'tcx> for ConstPropagator<'mir, 'tcx> {
             | TerminatorKind::GeneratorDrop
             | TerminatorKind::FalseEdges { .. }
             | TerminatorKind::FalseUnwind { .. } => {}
-            //FIXME(wesleywiser) Call does have Operands that could be const-propagated
-            TerminatorKind::Call { .. } => {}
+            TerminatorKind::Call { ref mut args, .. } => {
```

Comment:
I wonder if we should move to a general scheme where we `visit_operand` and mutate the operand directly instead of finding all the operands in all the terminators and statements and mutating them manually

Reply:
That would be specially useful for weirder `Operand` cases like `interpret::Immediate::ScalarPair` and `interpret::Indirect`

Reply:
In fact, in `ConstPropagator::replace_with_const` (https://github.com/rust-lang/rust/blob/9fc9dc99ee60a7ab603622ee83d14f8e05485740/src/librustc_mir/transform/const_prop.rs#L699) this exact thing is implemented already. There is even a comment about how `Scalar` and `ScalarPair` were relatively easy to implement compared to `Operand::Indirect`, and that because of that, `Indirect`'s const prop implementation was postponed.

I'm currently trying to adapt the `replace_with_const` code to work with a `ScalarPair` that came from an `Operand` (still trying to work out the types, because I don't have an `Rvalue` to capture my results). I think I 200% second your idea now @oli-obk, we're already having code duplication and it should be straight-forward-ish to do it, provided we work out the types so that indeed we can just directly mutate an `Operand` without needing things like `Rvalue`s to do it.

Reply:
@wesleywiser what do you think?

If we go ahead with this idea, I think I could do a separate PR to address it, since I feel that this requires quite some work to be done outside of `mir/transform/const_prop.rs`. I can work the `ScalarPair` cases et al there as well, since after this hypothetical function is implemented, expanding the current propagation at function callsites into `ScalarPair`s should be almost a one-liner.

Reply:
I think that's a great idea! My only concern is that `replace_with_const` has some special handling for `Rvalue::Aggregate` and we should make sure we don't regress the optimizations we do there.

https://github.com/rust-lang/rust/blob/82e90d64266b8a4b53935d629786e69610b33f25/src/librustc_mir/transform/const_prop.rs#L724-L730

Reply:
I've just realized that I left you hanging here. Yes, indeed. I wouldn't touch whatever optimizations are in there if at all possible.